File: backend/app/schemas/cash.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import SoftDeleteSchema, TimestampedSchema
from app.schemas.user import UserRead


class CashRegisterCreate(BaseModel):
    name: str = Field(min_length=2, max_length=100)
    location: str | None = Field(default=None, max_length=120)


class CashRegisterUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=100)
    location: str | None = Field(default=None, max_length=120)
    is_active: bool | None = None


class CashRegisterRead(SoftDeleteSchema):
    name: str
    location: str | None = None


class CashSessionOpen(BaseModel):
    cash_register_id: UUID
    opening_amount: Decimal = Decimal("0.00")


class CashReasonCatalogRead(BaseModel):
    reason_type: str
    code: str
    label: str
    description: str
    requires_notes: bool


class CashSessionCloseDenomination(BaseModel):
    denomination_value: Decimal = Field(gt=0)
    quantity: int = Field(ge=0)


class CashSessionClose(BaseModel):
    counted_amount: Decimal = Field(gt=0)
    denominations: list[CashSessionCloseDenomination] = Field(default_factory=list)
    observation: str | None = Field(default=None, max_length=500)
    supervisor_username: str | None = Field(default=None, min_length=3, max_length=50)
    supervisor_password: str | None = Field(default=None, min_length=8, max_length=128)


class CashSessionReopen(BaseModel):
    reason_code: str = Field(min_length=2, max_length=60)
    notes: str | None = Field(default=None, max_length=500)


class CashMovementCreate(BaseModel):
    movement_type: Literal["income", "withdrawal"]
    amount: Decimal = Field(gt=0)
    reason_code: str = Field(min_length=2, max_length=60)
    notes: str | None = Field(default=None, max_length=500)


class CashMovementRead(TimestampedSchema):
    cash_session_id: UUID
    created_by_user_id: UUID
    movement_type: str
    amount: Decimal
    reason: str
    notes: str | None = None
    created_by: UserRead


class CashSessionClosureDenominationRead(TimestampedSchema):
    cash_session_closure_id: UUID
    denomination_value: Decimal
    quantity: int
    subtotal: Decimal


class CashSessionClosureRead(TimestampedSchema):
    cash_session_id: UUID
    closed_by_user_id: UUID
    supervised_by_user_id: UUID | None = None
    expected_amount: Decimal
    counted_amount: Decimal
    difference_amount: Decimal
    observation: str | None = None
    reopened_at: datetime | None = None
    reopened_by_user_id: UUID | None = None
    reopen_reason: str | None = None
    closed_by: UserRead
    supervised_by: UserRead | None = None
    reopened_by: UserRead | None = None
    denominations: list[CashSessionClosureDenominationRead] = Field(default_factory=list)
    cash_register_name: str
    opened_by_name: str
    opened_at: datetime
    session_status: str


class CashSessionMetricsRead(BaseModel):
    sales_count: int
    completed_sales_count: int
    cancelled_sales_count: int
    total_sales_amount: Decimal
    cash_sales_amount: Decimal
    manual_income_amount: Decimal
    manual_withdrawal_amount: Decimal
    expected_cash_amount: Decimal


class CashSessionRead(TimestampedSchema):
    cash_register_id: UUID
    opened_by_user_id: UUID
    opening_amount: Decimal
    closing_amount: Decimal | None = None
    counted_amount: Decimal | None = None
    difference_amount: Decimal | None = None
    closing_observation: str | None = None
    status: str
    cash_register: CashRegisterRead
    opened_by: UserRead
    closed_by: UserRead | None = None
    supervised_by: UserRead | None = None
    movements: list[CashMovementRead] = Field(default_factory=list)


class CashSessionSummaryRead(CashSessionRead):
    metrics: CashSessionMetricsRead

File: frontend/src/features/cash/types/cash.ts
export type CashRegister = {
  id: string;
  name: string;
  location?: string | null;
  is_active: boolean;
};

export type CashReasonCatalogItem = {
  reason_type: "income" | "withdrawal" | "sale_cancellation" | "cash_reopen";
  code: string;
  label: string;
  description: string;
  requires_notes: boolean;
};

export type CashClosureDenomination = {
  id: string;
  cash_session_closure_id: string;
  denomination_value: string;
  quantity: number;
  subtotal: string;
  created_at: string;
  updated_at: string;
};

export type CashSessionClosure = {
  id: string;
  cash_session_id: string;
  closed_by_user_id: string;
  supervised_by_user_id?: string | null;
  expected_amount: string;
  counted_amount: string;
  difference_amount: string;
  observation?: string | null;
  reopened_at?: string | null;
  reopened_by_user_id?: string | null;
  reopen_reason?: string | null;
  created_at: string;
  updated_at: string;
  closed_by: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  };
  supervised_by?: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  } | null;
  reopened_by?: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  } | null;
  denominations: CashClosureDenomination[];
  cash_register_name: string;
  opened_by_name: string;
  opened_at: string;
  session_status: string;
};

export type CashSessionMetrics = {
  sales_count: number;
  completed_sales_count: number;
  cancelled_sales_count: number;
  total_sales_amount: string;
  cash_sales_amount: string;
  manual_income_amount: string;
  manual_withdrawal_amount: string;
  expected_cash_amount: string;
};

export type CashMovement = {
  id: string;
  cash_session_id: string;
  created_by_user_id: string;
  movement_type: "income" | "withdrawal";
  amount: string;
  reason: string;
  notes?: string | null;
  created_at: string;
  updated_at: string;
  created_by: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  };
};

export type CashSession = {
  id: string;
  cash_register_id: string;
  opened_by_user_id: string;
  opening_amount: string;
  closing_amount?: string | null;
  counted_amount?: string | null;
  difference_amount?: string | null;
  closing_observation?: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  cash_register: CashRegister;
  opened_by: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  };
  closed_by?: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  } | null;
  supervised_by?: {
    id: string;
    full_name: string;
    username: string;
    role: {
      id: string;
      name: string;
      permissions: string[];
    };
  } | null;
  movements: CashMovement[];
  metrics: CashSessionMetrics;
};

export type OpenCashSessionPayload = {
  cash_register_id: string;
  opening_amount: number;
};

export type CloseCashSessionPayload = {
  counted_amount: number;
  denominations?: Array<{
    denomination_value: number;
    quantity: number;
  }>;
  observation?: string;
  supervisor_username?: string;
  supervisor_password?: string;
};

export type CreateCashMovementPayload = {
  movement_type: "income" | "withdrawal";
  amount: number;
  reason_code: string;
  notes?: string;
};

export type ReopenCashSessionPayload = {
  reason_code: string;
  notes?: string;
};
